crates/walrus-indexer/src/async_task_manager.rs

```diff
@@ -0,0 +1,705 @@
+        })
+    }
+
+    /// The main retry queue processing loop.
+    async fn run(&mut self) {
+        loop {
+            self.start_processing_retry_tasks().await;
+
+            tokio::select! {
+                _ = self.shutdown_token.cancelled() => {
+                    tracing::info!("Retry handler shutting down");
+                    if let Some(handle) = self.task_handle.take() {
+                        handle.abort();
+                    }
+                    break;
+                }
+
+                // Listen for retry signals.
+                Some(_) = self.retry_notification_receiver.recv() => {
+                    let _guard = self.sync_mutex.lock().await;
+                    self.retry_notified = true;
+                }
+
+                // Check if current task finished.
+                _ = async {
+                    // Check if we have a handle without taking it
+                    let has_handle = {
+                        let _guard = self.sync_mutex.lock().await;
+                        self.task_handle.is_some()
+                    };
+
+                    if has_handle {
+                        // Wait for the task to complete
+                        if let Some(ref mut handle) = self.task_handle {
+                            handle.await.ok()
+                        } else {
+                            None
+                        }
+                    } else {
+                        // No task running, wait forever.
+                        std::future::pending::<Option<()>>().await
+                    }
+                } => {
+                    // Task finished, clear the handle.
+                    let _guard = self.sync_mutex.lock().await;
+                    self.task_handle = None;
+                }
+            }
+        }
+    }
+
+    /// Spawn a task that processes the retry queue once and exits.
+    async fn start_processing_retry_tasks(&mut self) {
+        let _guard = self.sync_mutex.lock().await;
+        if self.task_handle.is_some() {
+            return;
+        }
+        if !self.retry_notified {
+            return;
+        }
+
+        // Clear the retry_notified flag.
+        self.retry_notified = false;
+
+        let store = self.store.clone();
+        let executor = self.executor.clone();
+        let config = self.config.clone();
+        let shutdown_token = self.shutdown_token.clone();
+        let retry_notification_sender = self.retry_notification_sender.clone();
+        let semaphore = Arc::new(Semaphore::new(config.max_concurrent_retry_tasks));
+        let read_batch_size = config.read_batch_size;
+
+        let task_handle = tokio::spawn(async move {
+            Self::process_retry_queue(
+                store,
+                executor,
+                read_batch_size,
+                semaphore,
+                retry_notification_sender,
+                shutdown_token,
+            )
+            .await;
```

Comment:
suggestion: instead of calling `shutdown_token.is_cancelled()` in a few places inside the function, consider doing something like

```
        let task_handle = tokio::spawn(async move {
            select! {
                _ = shutdown_token.cancelled() => {
                    return;
                }
                _ = Self::process_retry_queue(
                    store,
                    executor,
                    read_batch_size,
                    semaphore,
                    retry_notification_sender,
                    shutdown_token.clone(),
                ) => {
                    return;
                }
            }
```